pkg/executor/join/hash_table_v2.go

```diff
@@ -0,0 +1,216 @@
+// Copyright 2024 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package join
+
+import (
+	"sync/atomic"
+	"unsafe"
+)
+
+type subTable struct {
+	rowData          *rowTable
+	hashTable        []unsafe.Pointer
+	posMask          uint64
+	isRowTableEmpty  bool
+	isHashTableEmpty bool
+}
+
+func (st *subTable) lookup(hashValue uint64) unsafe.Pointer {
+	return st.hashTable[hashValue&st.posMask]
+}
+
+func nextPowerOfTwo(value uint64) uint64 {
+	ret := uint64(2)
+	round := 1
+	for ; ret <= value && round <= 64; ret = ret << 1 {
+		round++
+	}
+	if round > 64 {
+		panic("input value is too large")
+	}
+	return ret
+}
+
+func newSubTable(table *rowTable) *subTable {
+	ret := &subTable{
+		rowData:          table,
+		isHashTableEmpty: false,
+		isRowTableEmpty:  false,
+	}
+	if table.rowCount() == 0 {
+		ret.isRowTableEmpty = true
+	}
+	if table.validKeyCount() == 0 {
+		ret.isHashTableEmpty = true
+	}
+	capacity := max(nextPowerOfTwo(table.validKeyCount()), uint64(1024))
```

Comment:
Maybe rename `capacity` to `hashTableLen` is better.

Reply:
Done